src/main/scala/abalone/variant/Variant.scala

```diff
@@ -46,75 +46,159 @@ abstract class Variant private[variant] (
+   you can easily generate the side moves as being the one before and the one after,
+   following a rotation :
+    \   /
+   -  o  -
+    /   \
+  e.g. if you are moving upRight the side moves to consider are "upLeft" and "right".
+
+  1. moves of 1 marble
+  2. generate any possible pair of marbles and use it to generate moves of 2 and 3 marbles
   then merge these as valid moves.
   */
-  def validMoves(situation: Situation): Map[Pos, List[Move]] = {
-    val movesOf1 = validMovesOf1(situation)
-    val lineMoves = movesOf1 ++ validLineMoves(situation).map {
-      case (k, v) => k -> (v ++ movesOf1.getOrElse(k, Iterable.empty))
-    }
-    val lineMovesAndPushes = lineMoves ++ validPushes(situation).map {
-      case (k, v) => k -> (v ++ lineMoves.getOrElse(k, Iterable.empty))
-    }
-    lineMovesAndPushes ++ validSideMoves(situation).map { // @TODO: this should reuse what was computed in validMovesOf1
-      case (k, v) => k -> (v ++ lineMovesAndPushes.getOrElse(k, Iterable.empty))
-    }
-  }
+  def validMoves(situation: Situation): Map[Pos, List[Move]] =
+    (validMovesOf1(situation).toList ++ validMovesOf2And3(situation).view.mapValues(_.map(_._2)).toList)
+      .groupBy(_._1).map{case(k, v) => k -> v.map(_._2).toSeq.flatten}.toMap
 
   def validMovesOf1(situation: Situation): Map[Pos, List[Move]] =
-    situation.board.piecesOf(situation.player).flatMap {
+    turnPieces(situation).flatMap {
       case ((pos, piece)) =>
         Map(pos ->
           pos.neighbours.flatten
           .filterNot(situation.board.pieces.contains(_))
-          .map(landingSquare => 
+          .map(landingSquare =>
             Move(piece, pos, landingSquare, situation, boardAfter(situation, pos, landingSquare), false)
           )
         )
     }.toMap
 
-  def validSideMoves(@nowarn situation: Situation):  Map[Pos, List[Move]] = {
-    Map()
-  }
+  def validMovesOf2And3(situation: Situation): Map[Pos, List[(String, Move)]] = {
+    val activePlayerPieces = situation.board.piecesOf(situation.player)
+    val opponentPieces = situation.board.piecesOf(!situation.player)
+
+    def createMove(orig: Pos, dest: Pos, category: String): (String, Move) =
+      (category, Move(Piece(situation.player, Role.defaultRole), orig, dest, situation, boardAfter(situation, orig, dest), true, if (category == "pushout") Some(dest) else None))
+
+    def generateSideMoves(lineOfMarbles: List[Pos], direction: Option[String]): List[(String, Move)] = {
+      def canMoveTowards(pos: Pos, dir: String): Boolean = {
+        situation.board.isEmptySquare(pos.dir(dir))
+      }
+
+      val possibleSideMovesDirections = Pos.sideMovesDirsFromDir(direction)
+      def possibleSideMoves: List[Option[(Pos, Pos)]] = possibleSideMovesDirections.map(
+        dir =>
+          if (lineOfMarbles.flatMap(
+            (pos) => Some(canMoveTowards(pos, dir))
+          ).contains(false)) None
+          else List(
```

Comment:
there's no need to use `List` here, you can just use a tuple and match on the tupe

Reply:
should be ok